src/Contract/Result/Warning.php

```diff
@@ -8,10 +8,13 @@
 
 /**
  * @psalm-immutable
+ *
+ * Represents a situation that, while valid, is not recommended. This can be
+ * used to guide the end-user to a more proper solution.
  */
 final class Warning implements Stringable
 {
-    private function __construct(private readonly string $message)
+    public function __construct(private readonly string $message)
```

Comment:
Why to you change the visibility of the constructor? I would prefer not to.

Reply:
This is to enable people writing extensions to create the class with a custom message. They might have their reason to create warning.

Reply:
What about adding an interface that people who write extensions could use. The `Warning` class is inside the contract namespace, therefore and `interface` makes sense, right? I think I mentioned to @dbrumann that I don't like to have concrete implementations inside the contract namespace, but I'm okay with default implementations.

Or would it be more useful to remove the final too, so everyone could decorate `Warning`?

I do don't like to write something like this:
```php
new Warning(sprintf('%s my message', $his->name))
```
if would prefer to be able to do something like this instead:
```php
MyWarning::causedForName($this->name);
```

Reply:
I see what you mean there. We do not have any standards about what should be an `interface`, what should be a concrete `class`, and what should be a `final class`. I propose this:
- For value objects (basically structs with optionally some helper functions):
 - If there are multiple implementations - make `interface` (like `TokenInterface`)
 - If they are created by deptrac internals - make them `final class` (like `FileOccurrence` or any of the triggered events)
 - If they are created by extension creator - make them non-final `class` so that they can be extended (like `Warning`)
- For the rest, aim for `interface` (like `ViolationCreatingInterface`)

Reply:
I spent a few hours going over the whole contract namespace, making incremental improvements here and there, and working with it, I came to a completely different conclusion now.

I propose the interface should be composed entirely of `interfaces` and `final class`. Nothing else. The final classes serve as inputs or outputs to our system and you cannot ever get anything else than those concrete implementations.

So for the situation above we could have:

```php
interface WarningInterface extends Stringable{...}
```

But I do not see any advantages to this. The difference:

```php
final class Warning implements Stringable
{
    public function __construct(private readonly string $message)
    {
    }
    public function __toString()
    {
        return $this->message;
    }
}

$a = new Warning(sprintf('%s my message', $this->name));

// or if you want
function forName(string $name): Warning {
 return new Warning(sprintf('%s my message', $name));
};
$a = forName($this-name);
```
vs
```php
interface WarningInterface extends Stringable{}
final class MyWarning implements WarningInterface
{
    public function __construct(private readonly string $message)
    {
    }
    public static function causedForName(string $name): self
    {
       return new self(sprintf('%s my message', $name));
    }
    
    public function __toString()
    {
        return $this->message;
    }
}

$a = MyWarning::causedForName($this-name);
```

Now I made you copy all the code that could have been part of our interface into your userland and forced you to be responsible for it. Made you write new classes just so that you can interact with deptrac. I don't see the value in that.

Reply:
Thank you for adding this extra mileage! I'm still not 100% convinced, but it's just a minor thing and we should focus on the more import stuff, getting the release out.